_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+smartphones or sensors, which collect gigantic amounts of data). In federated learning, each participant receives a copy of
+the model, trains on their own data, and sends back the gradients obtained to the central server, where gradients are averaged
+and applied to the model.
+
+This is good insofar as the data never leaves the individual devices; however, a lot of information can still be extracted
+from plain-text gradients. Imagine a smartphone app that provides trainable auto-completion for text messages. Even if
+gradient updates from many iterations are averaged, their distributions will greatly vary between individuals. Some form of
+encryption is needed. But then how is the server going to make sense of the encrypted gradients?
+
+One way to accomplish this relies on *secure multi-party computation* (SMPC).
+
+#### Secure multi-party computation
+
+In SMPC, we need a system of several agents who collaborate to provide a result no single agent could provide alone: "normal"
+computations (like addition, multiplication ...) on "secret" (encrypted) data. The assumption is that these agents are "honest
+but curious" -- honest, because they won't tamper with their share of data; curious in the sense that if they *were* (curious,
+that is), they wouldn't be able to inspect the data because it's encrypted.
+
+The principle behind this is *secret sharing*. A single piece of data -- a salary, say -- is "split up" into meaningless
+(hence, encrypted) parts which, when put together again, yield the original data. Here is an example.
+
+Say the parties involved are Julia, Greg, and me. The below function encrypts a single value, assigning to each of us their
+"meaningless" share:
+
+```{r}
+# a big prime number
+# all computations are performed in a finite field, for example, the integers modulo that prime
+Q <- 78090573363827
+ 
+encrypt <- function(x) {
+  # all but the very last share are random 
+  julias <- runif(1, min = -Q, max = Q)
+  gregs <- runif(1, min = -Q, max = Q)
+  mine <- (x - julias - gregs) %% Q
+  list (julias, gregs, mine)
+}
+
+# some top secret value no-one may get to see
+value <- 77777
+
+encrypted <- encrypt(value)
+encrypted
+```
+
+    [[1]]
+    [1] 7467283737857
+
+    [[2]]
+    [1] 36307804406429
+
+    [[3]]
+    [1] 34315485297318
+
+Once the three of us put our shares together, getting back the plain value is straightforward:
+
+```{r}
+decrypt <- function(shares) {
+  Reduce(sum, shares) %% Q  
+}
+
+decrypt(encrypted)
+```
+
+    77777
+
+As an example of how to compute on encrypted data, here's addition. (Other operations will be a lot less straightforward.) To
+add two numbers, just have everyone add their respective shares:
+
+```{r}
+add <- function(x, y) {
+  list(
+    # julia
+    (x[[1]] + y[[1]]) %% Q,
+    # greg
+    (x[[2]] + y[[2]]) %% Q,
+    # me
+    (x[[3]] + y[[3]]) %% Q
+  )
+}
+  
+x <- encrypt(11)
+y <- encrypt(122)
+
+decrypt(add(x, y))
+```
+
+    133
```

Comment:
What's the 133 here? If you evaluate the chunk, it should return the value on its own. I'm just looking at the Rmd right now, but, if you're getting this number twice, this would be why.